File: scripts/oneshot/02-build-source-pool.py
#!/usr/bin/env python3
"""
Phase 1 · Build source pool for the July 16 IP webinar campaign.

Python version — bypasses the node_modules / symlink mess on the Mac side.
Uses only stdlib + openpyxl (already in anaconda base). Talks to Neon via
its HTTP endpoint so no postgres driver needed.

Unions three sources into one deduped XLSX:
  - investors table (Neon)
  - crm_entries table (Neon)
  - anker-network-*.xlsx (LinkedIn export)

Dedupe order: linkedin_url → normalized(name+firm) → email.

Usage:
  NEON_DATABASE_URL='postgresql://...' \\
    python3 scripts/oneshot/02-build-source-pool.py \\
      ~/Downloads/anker-network-2026-07-11.xlsx
"""
import json, os, re, sys
from urllib import request as urlrequest
from urllib.parse import urlparse
from openpyxl import Workbook, load_workbook

# ---------- Neon HTTP ----------
NEON_URL = os.environ.get("NEON_DATABASE_URL")
if not NEON_URL:
    sys.exit("NEON_DATABASE_URL missing")

parsed = urlparse(NEON_URL)
NEON_HOST = parsed.hostname
NEON_ENDPOINT = f"https://{NEON_HOST}/sql"

def sql(query, params=None):
    """POST a SQL query to Neon's HTTP endpoint. Returns list of dict rows."""
    body = json.dumps({"query": query, "params": params or []}).encode("utf-8")
    req = urlrequest.Request(
        NEON_ENDPOINT,
        data=body,
        headers={
            "Neon-Connection-String": NEON_URL,
            "Neon-Raw-Text-Output": "true",
            "Neon-Array-Mode": "true",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    with urlrequest.urlopen(req, timeout=60) as r:
        data = json.loads(r.read())
    fields = [f["name"] for f in data.get("fields", [])]
    rows = data.get("rows", [])
    return [dict(zip(fields, row)) for row in rows]

# ---------- Input ----------
if len(sys.argv) < 2:
    sys.exit("Usage: python 02-build-source-pool.py <path/to/anker-network-*.xlsx>")
network_xlsx = os.path.expanduser(sys.argv[1])

def norm(s):
    return (s or "").strip().lower()

def norm_url(s):
    return re.sub(r"^https?://(www\.)?", "", norm(s)).rstrip("/")

# ---------- Pull investors ----------
print("→ Pulling investors from Neon…", flush=True)
investors = sql("""
    SELECT id::text, name, email, linkedin_url, title, firm, tags,
           COALESCE(metadata->>'lp_type', 'investor')  AS lp_type,
           COALESCE(metadata->>'sectors', '')          AS sectors,
           COALESCE(metadata->>'location', '')         AS location
    FROM investors
    WHERE name IS NOT NULL
""")
print(f"  {len(investors)} rows")

# ---------- Pull crm_entries ----------
print("→ Pulling crm_entries from Neon…", flush=True)
crm = sql("""
    SELECT id::text,
           COALESCE(display_name, linkedin_data->>'name') AS name,
           display_email                                  AS email,
           display_linkedin                               AS linkedin_url,
           linkedin_data->>'title'                        AS title,
           linkedin_data->>'company'                      AS firm,
           linkedin_data->>'headline'                     AS headline,
           stage
    FROM crm_entries
    WHERE display_name IS NOT NULL OR linkedin_data IS NOT NULL
""")
print(f"  {len(crm)} rows")

# ---------- Read network XLSX ----------
print(f"→ Reading LinkedIn network export: {network_xlsx}", flush=True)
wb = load_workbook(network_xlsx, read_only=True, data_only=True)
ws = wb.worksheets[0]
headers = [c.value for c in next(ws.iter_rows(min_row=1, max_row=1))]
idx = {h: i for i, h in enumerate(headers)}
def h(row, key):
    i = idx.get(key)
    return row[i] if i is not None else None

network = []
for row in ws.iter_rows(min_row=2, values_only=True):
    if not any(row):
        continue
    network.append({
        "name": h(row, "name"),
        "linkedin_url": h(row, "linkedin_url"),
        "headline": h(row, "headline"),
        "firm": h(row, "company"),
        "title": h(row, "title"),
        "location": h(row, "location"),
        "degree": h(row, "degree"),
        "previous_firm": h(row, "previous_company"),
        "previous_title": h(row, "previous_title"),
    })
wb.close()
print(f"  {len(network)} rows")

# ---------- Dedupe & merge ----------
print("→ Deduping (linkedin_url → name+firm → email)…", flush=True)
pool = {}
def upsert(row, source):
    key = None
    if row.get("linkedin_url"):
        key = f"li:{norm_url(row['linkedin_url'])}"
    elif row.get("name"):
        key = f"n:{norm(row['name'])}|f:{norm(row.get('firm'))}"
    elif row.get("email"):
        key = f"e:{norm(row['email'])}"
    else:
        return
    existing = pool.get(key, {})
    merged = {**existing}
    for k, v in row.items():
        if v:
            merged[k] = v
    srcs = set(existing.get("sources", []))
    srcs.add(source)
    merged["sources"] = sorted(srcs)
    pool[key] = merged

for r in investors: upsert(r, "neon-investors")
for r in crm:       upsert(r, "neon-crm")
for r in network:   upsert(r, "linkedin-export")

rows = list(pool.values())
print(f"  {len(rows)} unique contacts after dedupe", flush=True)

# ---------- Write source-pool.xlsx ----------
print("→ Writing source-pool.xlsx…", flush=True)
out = Workbook()
out_ws = out.active
out_ws.title = "Pool"
cols = ["name", "email", "linkedin_url", "firm", "title", "headline",
        "lp_type", "sectors", "location", "tags", "degree",
        "previous_firm", "previous_title", "sources", "stage", "id"]
out_ws.append(cols)
for r in rows:
    vals = []
    for c in cols:
        v = r.get(c)
        if isinstance(v, (list, tuple)):
            v = ", ".join(str(x) for x in v)
        vals.append(v if v is not None else "")
    out_ws.append(vals)
# Bold header
for cell in out_ws[1]:
    cell.font = cell.font.copy(bold=True)
out.save("source-pool.xlsx")

# ---------- Coverage report ----------
c_both = c_email = c_li = c_none = 0
for r in rows:
    e, l = bool(r.get("email")), bool(r.get("linkedin_url"))
    if e and l:   c_both += 1
    elif e:       c_email += 1
    elif l:       c_li += 1
    else:         c_none += 1
print(f"\n✓ source-pool.xlsx written ({len(rows)} rows)")
print(f"Coverage:")
print(f"  email + linkedin: {c_both}")
print(f"  email only:       {c_email}")
print(f"  linkedin only:    {c_li}")
print(f"  neither:          {c_none}")

File: scripts/oneshot/02-build-source-pool.mjs
/**
 * Phase 1 · Build source pool for the July 16 IP webinar campaign.
 *
 * Unions three sources into one deduped XLSX:
 *   - investors table (Neon)
 *   - crm_entries table (Neon)
 *   - anker-network-*.xlsx (LinkedIn export)
 *
 * Dedupe order: linkedin_url → normalized(name+firm) → email.
 *
 * Usage:
 *   NEON_DATABASE_URL='postgresql://…' \
 *     node scripts/oneshot/02-build-source-pool.mjs \
 *       ~/Downloads/anker-network-2026-07-11.xlsx
 *
 * Output: source-pool.xlsx in cwd. Every row has a `source` column so we
 * know where the contact came from (neon-investors, neon-crm, linkedin-export).
 */
import { neon } from "@neondatabase/serverless"
import ExcelJS from "exceljs"
import { readFile } from "node:fs/promises"

const url = process.env.NEON_DATABASE_URL
if (!url) { console.error("NEON_DATABASE_URL missing"); process.exit(1) }
const networkXlsx = process.argv[2]
if (!networkXlsx) { console.error("Usage: node 02-build-source-pool.mjs <path/to/anker-network-*.xlsx>"); process.exit(1) }

const sql = neon(url)

const norm = (s) => (s || "").toString().trim().toLowerCase().replace(/\s+/g, " ")
const normUrl = (s) => norm(s).replace(/^https?:\/\/(www\.)?/, "").replace(/\/$/, "")

console.log("→ Pulling investors from Neon…")
const investors = await sql`
  SELECT id::text, name, email, linkedin_url, title, firm, tags,
         COALESCE(metadata->>'lp_type', 'investor')  AS lp_type,
         COALESCE(metadata->>'sectors', '')         AS sectors,
         COALESCE(metadata->>'location', '')        AS location
  FROM investors
  WHERE name IS NOT NULL
`
console.log(`  ${investors.length} rows`)

console.log("→ Pulling crm_entries from Neon…")
const crmRows = await sql`
  SELECT id::text,
         COALESCE(display_name, linkedin_data->>'name') AS name,
         display_email                                  AS email,
         display_linkedin                               AS linkedin_url,
         linkedin_data->>'title'                        AS title,
         linkedin_data->>'company'                      AS firm,
         linkedin_data->>'headline'                     AS headline,
         stage
  FROM crm_entries
  WHERE display_name IS NOT NULL OR linkedin_data IS NOT NULL
`
console.log(`  ${crmRows.length} rows`)

console.log(`→ Reading LinkedIn network export: ${networkXlsx}`)
const wb = new ExcelJS.Workbook()
await wb.xlsx.readFile(networkXlsx)
const ws = wb.worksheets[0]
const headers = ws.getRow(1).values.map((v) => (v ? v.toString().trim() : ""))
const idx = (name) => headers.indexOf(name)
const networkRows = []
ws.eachRow({ includeEmpty: false }, (row, n) => {
  if (n === 1) return
  const v = row.values
  networkRows.push({
    source: "linkedin-export",
    name: v[idx("name")] || null,
    linkedin_url: v[idx("linkedin_url")] || null,
    headline: v[idx("headline")] || null,
    firm: v[idx("company")] || null,
    title: v[idx("title")] || null,
    location: v[idx("location")] || null,
    degree: v[idx("degree")] || null,
    previous_firm: v[idx("previous_company")] || null,
    previous_title: v[idx("previous_title")] || null,
  })
})
console.log(`  ${networkRows.length} rows`)

console.log("→ Deduping (linkedin_url → name+firm → email)…")
const pool = new Map()  // key → merged row
const upsert = (row, source) => {
  const key1 = row.linkedin_url ? `li:${normUrl(row.linkedin_url)}` : null
  const key2 = row.name ? `n:${norm(row.name)}|f:${norm(row.firm)}` : null
  const key3 = row.email ? `e:${norm(row.email)}` : null
  const key = key1 || key2 || key3 || `x:${Math.random()}`
  const existing = pool.get(key) || {}
  const merged = {
    ...existing,
    ...Object.fromEntries(Object.entries(row).filter(([, v]) => v)),
    sources: Array.from(new Set([...(existing.sources || []), source])),
  }
  pool.set(key, merged)
}
for (const r of investors) upsert(r, "neon-investors")
for (const r of crmRows) upsert(r, "neon-crm")
for (const r of networkRows) upsert(r, "linkedin-export")

const rows = Array.from(pool.values())
console.log(`  ${rows.length} unique contacts after dedupe`)

// Write out
console.log("→ Writing source-pool.xlsx…")
const out = new ExcelJS.Workbook()
const outWs = out.addWorksheet("Pool")
const cols = ["name", "email", "linkedin_url", "firm", "title", "headline",
              "lp_type", "sectors", "location", "tags", "degree",
              "previous_firm", "previous_title", "sources", "stage"]
outWs.addRow(cols)
for (const r of rows) {
  outWs.addRow(cols.map((c) => {
    const v = r[c]
    if (Array.isArray(v)) return v.join(", ")
    return v ?? ""
  }))
}
outWs.getRow(1).font = { bold: true }
await out.xlsx.writeFile("source-pool.xlsx")
console.log(`✓ source-pool.xlsx written (${rows.length} rows)`)

// Summary of coverage
const counts = { withEmail: 0, withLinkedin: 0, withBoth: 0, withNeither: 0 }
for (const r of rows) {
  const hasE = !!r.email
  const hasL = !!r.linkedin_url
  if (hasE && hasL) counts.withBoth++
  else if (hasE) counts.withEmail++
  else if (hasL) counts.withLinkedin++
  else counts.withNeither++
}
console.log("\nCoverage:")
console.log(`  email + linkedin: ${counts.withBoth}`)
console.log(`  email only:       ${counts.withEmail}`)
console.log(`  linkedin only:    ${counts.withLinkedin}`)
console.log(`  neither:          ${counts.withNeither}`)
